db/users.go

```diff
@@ -220,6 +220,10 @@ func (dbc *DatabaseContext) UpdatePrincipal(ctx context.Context, updates *auth.P
 		// On cas error, retry.  Otherwise break out of loop
 		if base.IsCasMismatch(err) {
 			base.InfofCtx(ctx, base.KeyAuth, "CAS mismatch updating principal %s - will retry", base.UD(princ.Name()))
+			// release the sequence number we allocated in the failed update to avoid an abandoned sequence
+			if err := dbc.sequences.releaseSequence(ctx, nextSeq); err != nil {
+				base.InfofCtx(ctx, base.KeyAuth, "Error releasing unused sequence %d after CAS retry: %v", nextSeq, err)
```

Comment:
I think I might add the principal name here so you can correlate with the above message.